epaint/src/shape.rs

```diff
@@ -464,3 +479,902 @@ fn dashes_from_line(
+    pub fn from_points_stroke(points: Vec<Pos2>, stroke: impl Into<Stroke>) -> Self {
+        crate::epaint_assert!(
+            points.len() == 3,
+            "Quadratic needs 3 points"
+        );
+
+        QuadraticBezierShape {
+            points: points.try_into().unwrap(), // it's safe to unwrap because we just checked
+            closed: false,
+            fill: Default::default(),
+            stroke: stroke.into(),
+        }
+    }
+    
+    /// bounding box of the quadratic bezier shape
+    pub fn bounding_rect(&self) -> Rect {
+        let (mut min_x, mut max_x) = if self.points[0].x < self.points[2].x {
+            (self.points[0].x, self.points[2].x)
+        } else {
+            (self.points[2].x, self.points[0].x)
+        };
+        let (mut min_y, mut max_y) = if self.points[0].y < self.points[2].y {
+            (self.points[0].y, self.points[2].y)
+        } else {
+            (self.points[2].y, self.points[0].y)
+        };
+
+        quadratic_for_each_local_extremum(self.points[0].x, self.points[1].x, self.points[2].x, &mut |t|{
+            let x = self.sample(t).x;
+            if x < min_x {
+                min_x = x;
+            }
+            if x > max_x {
+                max_x = x;
+            }
+        });
+
+        quadratic_for_each_local_extremum(self.points[0].y, self.points[1].y, self.points[2].y, &mut |t|{
+            let y = self.sample(t).y;
+            if y < min_y {
+                min_y = y;
+            }
+            if y > max_y {
+                max_y = y;
+            }
+        });
+        
+        Rect {
+            min: Pos2 { x: min_x, y: min_y },
+            max: Pos2 { x: max_x, y: max_y },
+        }
+    }
+
+    /// Calculate the point (x,y) at t based on the quadratic bezier curve equation.
+    /// t is in [0.0,1.0]
+    /// https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Quadratic_B.C3.A9zier_curves
+    /// 
+    pub fn sample(&self, t: f32) -> Pos2 {
+        crate::epaint_assert!(
+            t >= 0.0 && t <= 1.0,
+            "the sample value should be in [0.0,1.0]"
+        );
+
+        let h = 1.0 - t;
+        let a = t * t;
+        let b = 2.0 * t * h;
+        let c = h * h;
+        let result = self.points[2].to_vec2() * a
+            + self.points[1].to_vec2() * b
+            + self.points[0].to_vec2() * c;
+        result.to_pos2()
+    }
+
+    /// find a set of points that approximate the quadratic bezier curve.
+    /// the number of points is determined by the tolerance.
+    /// the points may not be evenly distributed in the range [0.0,1.0] (t value)
+    pub fn flatten(&self, tolerance:Option<f32>)->Vec<Pos2>{
+        let tolerance =tolerance.unwrap_or( (self.points[0].x-self.points[2].x).abs()*0.001);
+        let mut result = Vec::new();
+        self.for_each_flattened_with_t(tolerance, &mut |p,_t|{
+            result.push(p);
+        });
+        result
+    }
+
+    // copied from https://docs.rs/lyon_geom/latest/lyon_geom/
```

Comment:
There is a lot of code copied from lyon. I think we should add a credit to it in `README.md` under the `## Credits` section